lib/spack/spack/caches.py

```diff
@@ -59,24 +60,31 @@ def _fetch_cache():
 
 class MirrorCache(object):
     def __init__(self, root, skip_unstable_versions):
-        self.root = os.path.abspath(root)
+        self.root = spack.util.url.parse(root)
         self.skip_unstable_versions = skip_unstable_versions
 
     def store(self, fetcher, relative_dest):
         """Fetch and relocate the fetcher's target into our mirror cache."""
 
         # Note this will archive package sources even if they would not
         # normally be cached (e.g. the current tip of an hg/git branch)
-        dst = os.path.join(self.root, relative_dest)
-        mkdirp(os.path.dirname(dst))
+        dst = spack.util.url.join(self.root, relative_dest)
+        local_dst = spack.util.url.local_file_path(dst)
+        if local_dst:
+            mkdirp(os.path.dirname(local_dst))
         fetcher.archive(dst)
```

Comment:
@scheibelp 

I'm debugging a set of test failures with my changes.  I've narrowed my search down to these particular changes.  I made them without knowing if they're really appropriate, and I would greatly appreciate a generous amount of scrutiny, here.

My main questions concern the `fetcher.archive(dst)` call.  I observe that this call fails in testing (`spack unit-test -k all_mirror`) because the `dst` variable is not a file path, but a URL (due to the changes), and `fetcher.archive()` doesn't work with URLs.  The fetcher is a `VCSFetchStrategy`.

What do you think is the right change to make here?  Given that `dst` is a URL now (that can refer to remote resources), what would a URL-aware version of `fetcher.archive(dst)` look like?  Does `VCSFetchStrategy` also need to be URL-aware?  Do all the fetch strategies need to be?

I worry that either there is a disconnect between the assumptions made by the `MirrorCache` class and the various fetchers, or something is fundamentally wrong with my approach.  Would greatly appreciate your thoughts.

Reply:
After some more digging, I see that there is already precedent for fetchers being aware that their `destinations` are (possibly remote) URLs.  See `URLFetchStrategy.archive()`.

I'm pretty sure this is what @robo-wylder was originally trying to point out with his earlier comment.  I reproduce those comments below:

> I'm not sure this is necessary, the fetcher.archive() call itself seems to do the right thing without having to handle remote cases specially. Local directories get created automatically for local paths and for remote paths the fetcher knows how to push the file that already exists locally to the url. Look at the archive() method for the URLFetchStrategy and the implementation of push_to_url() in web.py.
> 
> I think the issue before was that this always forced a local path instead of just attempting to store the file in dst directly. Did the tests fail without the extra logic to localize the path?

As we can see, however, this is not the case for all fetchers (The `VCSFetchStrategy` still assumes local file path).  This leads me to believe that what needs to be done is to go through all of the fetch strategies and update them to handle the case where their fetch destination is a URL.  Considering that this would be a large change, I think more discussion around it is warranted.

Again, very interested in hearing what you think, @scheibelp.

Reply:
> what would a URL-aware version of fetcher.archive(dst) look like? 

I'm thinking you can copy the `URLFetchStrategy.archive` function's use of `web_util.push_to_url`: Right now all the `VCSFetchStrategy`s need to create their own archive, so I think instead of using `tar czf` with the mirror as the destination, you will want to create a temp dir with the archive, then `web_util.push_to_url` the resulting file in the temp directory.

> Does VCSFetchStrategy also need to be URL-aware? Do all the fetch strategies need to be?

Yes and yes. Overall though I think you should only have to update one function: `VcsFetchStrategy.archive()` - I don't see anything else implementing `archive()`.

If that doesn't clarify what needs to be done (or it seems like I overlooked something) let me know.